File: LeetCode/70.climbing-stairs.cpp
/*
 * @lc app=leetcode id=70 lang=cpp
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution
{
private:
    unordered_map<int, int> m;

public:
    int climbStairs(int n)
    {
        if (m.find(n) != m.end())
        {
            return m[n];
        }
        if (n <= 2)
        {
            return n;
        }
        int result = climbStairs(n - 1) + climbStairs(n - 2);
        m[n] = result;
        return m[n];
    }
};
// @lc code=end

File: LeetCode/70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#


# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 0 or n == 1:
            return 1
        prev, curr = 1, 1

        for _ in range(2, n + 1):
            temp = curr
            curr = prev + curr
            prev = temp

        return curr


# @lc code=end
